### ISSUE TITLE
Inconsistency with connect/login 

### ISSUE THREAD

**Issue by nixtux** [spring/uberserver#295]
Players can't seem to login again.

**Comment by nixtux** [comment]
Login works again.

**Comment by nixtux** [comment]
Closed too soon still valid

**Comment by silentwings** [comment]
There is a bug (longstanding) causing this that atm we don't know how fix, although I think there is a chance that some changes currently only in dev versions of the lobbyserver will shed some light on it (by making cleans of dbs more verbose).

**Comment by abma** [comment]
multiple CLEANUP always fixed the issue. RELOAD does nothing as it doesn't touch "data".

**Comment by silentwings** [comment]
i'll test again, but i'm pretty sure RELOAD was needed (all 3 times) before hosts were able to connect again

importlib.reload *does* touch data - from the docs on https://docs.python.org/3.4/library/importlib.html

"Python module’s code is recompiled and the module-level code re-executed, defining a new set of objects which are bound to names in the module’s dictionary by reusing the loader which originally loaded the module. "

so it could reset the values of global variables, for example

**Comment by abma** [comment]
```
2018-12-08 16:21:36 ERROR Protocol.cleanup:2938  missing user:  False session_id: 56920
2018-12-08 16:21:36 ERROR Protocol.cleanup:2933  duplicate user:
2018-12-08 16:21:36 ERROR Protocol.cleanup:2933  duplicate user: ZafirPL
2018-12-08 16:21:36 ERROR Protocol.cleanup:2938  missing user: Mercury True session_id: 52925
2018-12-08 16:21:36 ERROR Protocol.cleanup:2938  missing user: Deon True session_id: 52947
2018-12-08 16:21:36 ERROR Protocol.cleanup:2946  deleted invalid session: 56920
2018-12-08 16:21:36 ERROR Protocol.cleanup:2946  deleted invalid session: 56955
2018-12-08 16:21:36 ERROR Protocol.cleanup:2946  deleted invalid session: 52897 ZafirPL
2018-12-08 16:21:36 ERROR Protocol.cleanup:2946  deleted invalid session: 52925 Mercury
2018-12-08 16:21:36 ERROR Protocol.cleanup:2946  deleted invalid session: 52947 Deon
2018-12-08 16:21:37 INFO  Protocol._SendLoginInfo:958  [56955] <[ACE]Synchronisme> logged in (access=user).
2018-12-08 16:21:37 INFO  Protocol._checkCompat:496  [56955] <[ACE]Synchronisme> client "SPADS v0.12.4" missing compat flags: p tls
2018-12-08 16:21:37 INFO  Protocol._remove:230  [56965] <[ACE]Sure> disconnected from 78.199.212.117: Connection was closed cleanly.
2018-12-08 16:21:37 INFO  Protocol._remove:230  [56968] <[ACE]Synchronous> disconnected from 78.199.212.117: Connection to the other side was lost in a non-clean fashion.
2018-12-08 16:21:37 INFO  Protocol._new:200  [56976] Client connected from 88.124.248.198:52111
2018-12-08 16:21:38 INFO  Protocol._remove:230  [56959] <[ACE]Synchrone> disconnected from 78.199.212.117: Connection was closed cleanly.
2018-12-08 16:21:38 ERROR DataHandler.broadcast:371  Traceback (most recent call last):
  File "/home/lobby/uberserver/DataHandler.py", line 369, in broadcast
    self.multicast(channel.users, msg, ignore, sourceClient)
  File "/home/lobby/uberserver/DataHandler.py", line 344, in multicast
    if not client.logged_in:
AttributeError: 'NoneType' object has no attribute 'logged_in'
 
2018-12-08 16:21:38 ERROR Protocol.clientFromSession:622  Couldn't get client from session_id: 56955 NoneType: None
2018-12-08 16:21:38 ERROR Protocol.cleanup:2920  deleting session id <56955> from channel [ACE]
2018-12-08 16:21:38 ERROR Protocol.cleanup:2938  missing user:  False session_id: 56976
2018-12-08 16:21:38 ERROR Protocol.cleanup:2946  deleted invalid session: 56976
2018-12-08 16:21:38 ERROR Protocol.cleanup:2952  missing session for [ACE]Synchronisme
2018-12-08 16:21:38 ERROR Protocol.cleanup:2960  Deleted username without session: [ACE]Synchronisme
2018-12-08 16:21:38 INFO  Protocol._remove:230  [56955] <[ACE]Synchronisme> disconnected from 78.199.212.117: name 'client' is not defined

```

**Comment by silentwings** [comment]
Having a pair of "duplicate" users with the same blank-string username is probably not worrying; perhaps they sent STLS but didn't LOGIN/REGISTER yet. The other parts looks like genuine issues.

**Comment by silentwings** [comment]
I think I'd like to wait until the live server is updated before trying to debug this because
(1) I made the db side of cleanup log what it was doing.
(2) I couldn't commit a fix anyway (without making a new github account) until #282 is dealt with.
(3) I think we'll need to log more info, to find out how/when the duplicates are created, before we solve this one. I haven't yet understood why (or, if) duplicate users/sessions are what prevents autohosts from working.

**Comment by silentwings** [comment]
Ok! I'll have a proper go at solving this one soon.

**Comment by silentwings** [comment]
Another stacktrace of likely same issue:

2019-02-16 16:58:09 INFO  Protocol._SendLoginInfo:1088  [8759] <murtagon> logged in (access=user).
2019-02-16 16:58:09 ERROR Protocol.clientFromSession:711  Couldn't get client from session_id: 679 NoneType: None

2019-02-16 16:58:09 ERROR Protocol.cleanup:2965  deleting battle <Battle.Battle object at 0x7f3a24a27390>
2019-02-16 16:58:09 ERROR twistedserver.dataReceived:56  Error in handling data from client: murtagon 74.77.21.170 45348 <Battle.Battle object at 0x7f3a24a27390>, b'#2 LOGIN murtagon ************************ 0 192.168.0.100 SpringLobby 0.267 (unix x64)\t553479703\ta m sp cl p\n', Traceback (most recent call last):
  File "/home/lobby/uberserver/twistedserver.py", line 52, in dataReceived
    self.Handle(data.decode("utf-8"))
  File "/home/lobby/uberserver/Client.py", line 161, in Handle
    self.HandleProtocolCommands(self.data.split("\n"), flood_limits)
  File "/home/lobby/uberserver/Client.py", line 197, in HandleProtocolCommands
    self.HandleProtocolCommand(command)
  File "/home/lobby/uberserver/Client.py", line 167, in HandleProtocolCommand
    self._root.protocol._handle(self, cmd)
  File "protocol/Protocol.py", line 413, in _handle
    function(*([client] + fun_args))
  File "protocol/Protocol.py", line 1078, in in_LOGIN
    self._SendLoginInfo(client)
  File "protocol/Protocol.py", line 1103, in _SendLoginInfo
    client.RealSend(self.client_AddBattle(client, battle))
  File "protocol/Protocol.py", line 789, in client_AddBattle
    host = self.clientFromSession(battle.host)
  File "protocol/Protocol.py", line 712, in clientFromSession
    self.cleanup()
  File "protocol/Protocol.py", line 2966, in cleanup
    del self._root.battles[battle]
KeyError: <Battle.Battle object at 0x7f3a24a27390>

**Comment by silentwings** [comment]
happened again - somehow a user managed to login twice
cleanup "worked" but needed fixing so log was not useful https://github.com/spring/uberserver/commit/c31095c39aa551a02c65b5c66445a5c506f8cc6b

**Comment by silentwings** [comment]
2019-03-24 22:34:07 INFO  Protocol._new:275  [946] Client connected from ***********
2019-03-24 22:34:07 INFO  Protocol._SendLoginInfo:1058  [946] <[PinK]pta[Q___Q]> logged in (access=user).
2019-03-24 22:34:07 INFO  Protocol._checkCompat:259  [946] <[PinK]pta[Q___Q]> client "SpringLobby 0.264 (win x32)" sent incorrect compat flags {'a', 'm', 'cl', 'p', 'sp'} -- missing: u l t, deprecated: a m p, unknown:

caused slew of autohost dc/s and

2019-03-24 22:45:11 ERROR Protocol.cleanup:2837  client with missing username: [PinK]pta[Q___Q] 946
2019-03-24 22:45:11 ERROR Protocol.cleanup:2846  deleted invalid client: [PinK]pta[Q___Q] 946
2019-03-24 22:45:11 ERROR Protocol.cleanup:3009  deleted invalid session_id 946 from channel newbies
2019-03-24 22:45:11 ERROR Protocol.cleanup:3009  deleted invalid session_id 946 from channel ba
2019-03-24 22:45:11 ERROR Protocol.cleanup:3009  deleted invalid session_id 946 from channel main
2019-03-24 22:45:11 ERROR Protocol.cleanup:3021  deleted empty channel king
2019-03-24 22:45:11 ERROR Protocol.cleanup:3009  deleted invalid session_id 946 from channel pl
2019-03-24 22:45:11 INFO  Protocol.cleanup:3035  deleted:
 1 clients, 0 usernames, 0 user_ids
 0 bridged_locations, 0 bridged_usernames, 0 bridged_user_ids, 0 bridge_external_ids, 0 bridge_locations
 0 battles, 0 battle_users, 0 battle_pending_users
 1 channels, 4 channel_users, 0 channel_bridged_users
 found 0 mismatches

**Comment by silentwings** [comment]
Hopefully fixed (rel #335, https://github.com/spring/uberserver/commit/1fa91726789dde1dd17b55f965774782c5716764)
